test: modify weight estimate in functional tests

m3dwards opened the pull request:
Fixes: https://github.com/bitcoin/bitcoin/issues/25164

The wallet_send functional test has been flaky due to a slightly overestimated weight calculation. This PR makes the weight calculation more accurate, although occasionally, due to how ECDSA signatures can be different lengths it might slightly over estimate. The assertion in the test can handle this slight variation and so should continue passing.

Update:

Because the signature can be shorter that is used in the weight estimation or the final transaction the estimate could be both slightly smaller or slightly larger.

achow101 commented on test/functional/wallet_send.py:
```diff
@@ -543,11 +543,16 @@ def run_test(self):
         # Calculate the input weight
-        # (prevout + sequence + length of scriptSig + scriptsig + 1 byte buffer) * WITNESS_SCALE_FACTOR + num scriptWitness stack items + (length of stack item + stack item) * N stack items + 1 byte buffer
+        # (prevout + sequence + length of scriptSig + scriptsig) * WITNESS_SCALE_FACTOR + len of num scriptWitness stack items + (length of stack item + stack item) * N stack items
+        # Note that occasionally this weight estimate may be slightly larger than the real weight
```
The reverse is also true - the estimate can be a bit smaller than the real weight because the tx that we are using to perform this size calculation could have a shorter signature. So I don't think this actually fixes the problem, it just results in a fee too low error occasionally rather than a fee too high error.

m3dwards replied:
The `assert_fee_amount` function allows the estimate to be a bit larger with a 2 vbyte buffer but not smaller.

I can think of a few options:

- Replicate the 2 vbyte allowance in the smaller direction
- As you already suggested switch the test to taproot

I'm leaning towards the latter as I don't like the idea of adding an allowance for every test that uses that assert.

As an aside I think it would be better to have two fee assertion functions, one that is precise and another that takes an allowance that can be specified in each tests. This way the intention of the test is clear, at the moment tests benefit from this allowance whether they need it or not.

@achow101 what are your thoughts?

achow101 replied:
Actually, using taproot might result in a little bit less test coverage. IIRC the script chosen for the external utxo in this test is p2sh-segwit because it will have both a scriptSig and scriptWitness. This lets us check that coin control is applying the weight correctly for both legacy and segwit scripts. As Taproot is segwit only, we'd lose the coverage of the legacy side of things.

A different solution we could try is to check that the feerate is as we expected, rather than checking that the fee was expected. Since we are targeting an integer sat/vb feerate, the calculated feerate of the final tx should only ever be higher than the target by less than 1 sat/vb, so we could just calculate the sat/vbyte feerate, rounded down, and check that was what we expected.

m3dwards replied:
If the final feerate is only ever higher than the expected then the existing assertion should take care of that.

But I understood that the estimate can both be high or low depending on if a signature used for the estimate was smaller or the signature used in the final transaction was smaller.

We can still calculate the feerate in sats/vb but rounding up or down to the nearest integer.

achow101 reviewed (commented):
There is a test in rpc_psbt.py that does basically the same thing. Any changes made here should probably be made to that test as well since it tests the same feature in the psbt workflow.

m3dwards committed:
test: fix weight estimates in functional tests

Updates the weight estimate to be more accurate by removing byte buffers and calculating the length of the count of scriptWitnesses rather than just using the count itself.

m3dwards committed:
test: fix flaky wallet_send functional test

Rather than asserting that the exact fees are the same, check the fee rate rounded to nearest interger. This will account for small differences in fees caused by variability in ECDSA signature lengths.

m3dwards commented:
Force pushed new version addressing @achow101 's comments.

achow101 commented:
ACK e67ab174c9c04bba7a10724b7e694dd57f732177

S3RK commented on test/functional/wallet_send.py:
```diff
@@ -543,11 +543,16 @@ def run_test(self):
-        len_scriptwitness = (sum([(len(x) // 2) + len(ser_compact_size(len(x) // 2)) for x in psbt_in["final_scriptwitness"]]) + len(psbt_in["final_scriptwitness"]) + 1) if "final_scriptwitness" in psbt_in else 0
-        input_weight = ((40 + len_scriptsig) * WITNESS_SCALE_FACTOR) + len_scriptwitness
+        len_scriptsig += len(ser_compact_size(len_scriptsig))
+        len_scriptwitness = (sum([(len(x) // 2) + len(ser_compact_size(len(x) // 2)) for x in psbt_in["final_scriptwitness"]]) + len(ser_compact_size(len(psbt_in["final_scriptwitness"])))) if "final_scriptwitness" in psbt_in else 0
```
Should we also divide `len(psbt_in["final_scriptwitness"])` by 2 to get size in bytes?

m3dwards replied:
The divide by 2 is for turning a hex string into a count of bytes.

The length (read count) of the scriptWitnesses is just that, a count and not a hex string. Further this count is passed into function `ser_compact_size` that returns bytes so a simple length of this gives us the byte length we are after.

S3RK replied:
Yes, you're right. I thought witness is also serialised. Got myself confused.

S3RK commented:
Code review ACK e67ab174c9c04bba7a10724b7e694dd57f732177
